### PR TITLE
check_oom_events initial commit

### PR DESCRIPTION
**Pull request by oktopuz**
internal ticket PAASTA-567

This is the very first version of the check that I'm going to use for dogfooding.
Therefore no docstrings, no tests.

### CODE REVIEW

File: paasta_tools/check_oom_events.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        instance.service,
+        instance.instance,
+    )
+    monitoring_overrides = instance.get_monitoring()
+    monitoring_overrides['page'] = False
+    monitoring_overrides['ticket'] = False
+    monitoring_overrides['alert_after'] = '1m'
+    monitoring_overrides['runbook'] = ['http://y/none']  # TODO: needs a link
+    monitoring_overrides['realert_every'] = args.realert_every
+    monitoring_overrides['tip'] = 'Increase memory limit.'
```
Comment (solarkennedy) [inline paasta_tools/check_oom_events.py]: "Try bumping the memory limit past $instance.get_mem()"

File: paasta_tools/check_oom_events.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    if monitoring_overrides['team'] == 'operations':  # TODO: remove after initial testing
```
Comment (solarkennedy) [inline paasta_tools/check_oom_events.py]: Can you send all alerts to yourself instead by overriding the team directly? (to noop)

File: paasta_tools/check_oom_events.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        )
+
+
+def main(sys_argv):
+    args = parse_args(sys_argv[1:])
+    cluster = load_system_paasta_config().get_cluster()
+    victims = latest_oom_events(cluster, args.superregion)
+    for s_i in get_services_for_cluster(cluster, soa_dir=args.soa_dir):
+        instance = get_instance_config(
+            s_i[0], s_i[1], cluster,
```
Comment (solarkennedy) [inline paasta_tools/check_oom_events.py]: kwags

File: paasta_tools/check_oom_events.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            output=status[1],
+            soa_dir=instance.soa_dir,
+        )
+
+
+def main(sys_argv):
+    args = parse_args(sys_argv[1:])
+    cluster = load_system_paasta_config().get_cluster()
+    victims = latest_oom_events(cluster, args.superregion)
+    for s_i in get_services_for_cluster(cluster, soa_dir=args.soa_dir):
```
Comment (solarkennedy) [inline paasta_tools/check_oom_events.py]: Can this be

    for service, instance in get_ser...

?

File: paasta_tools/check_oom_events.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    return res
+
+
+def compose_sensu_status(service_instance, events):
+    """
+    :param service_instance: a tuple (service, instance)
+    :param events: a list of tuples (hostname, container_id, process_name)
+    """
+    if len(events) == 0:
+        return (Status.OK, 'oom-killer is calm.')
```
Comment (solarkennedy) [inline paasta_tools/check_oom_events.py]: No oom events for $service.$instance in the last N minutes

File: paasta_tools/check_oom_events.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    elif len(events) == 1:
```
Comment (solarkennedy) [inline paasta_tools/check_oom_events.py]: I don't think you need the special case here, I think the generic one in the `else` works just fine for 1

File: paasta_tools/check_oom_events.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        return (Status.CRITICAL, 'killing 1 process/min (%s).' % events[0][2])
+    else:
+        return (
+            Status.CRITICAL, 'killing %d processes/min (%s).' %
```
Comment (solarkennedy) [inline paasta_tools/check_oom_events.py]: "The Out Of Memory killer killed %d processes in the last n minutes in $service.$instance containers.

Here are the events:
%...

OOM killed things usually do not log much, but you can get some logs with:

    paasta logs ....

File: paasta_tools/check_oom_events.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        help="The superregion to read OOM events from.",
+    )
+    return parser.parse_args(args)
+
+
+def oom_events(cluster, superregion, num_lines=1000):
+    """Read the latest 'num_lines' lines from OOM_EVENTS_STREAM and iterate over them."""
+    host_port = choice(scribereader.get_default_scribe_hosts(tail=True))
+    stream = scribereader.get_stream_tailer(
+        OOM_EVENTS_STREAM, host_port['host'], host_port['port'], True,
```
Comment (solarkennedy) [inline paasta_tools/check_oom_events.py]: kwargs please

File: paasta_tools/check_oom_events.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+            j = json.loads(line)
+            if j.get('cluster', '') == cluster:
+                yield j
+        except json.decoder.JSONDecodeError:
+            pass
+
+
+def latest_oom_events(cluster, superregion, interval=60):
+    """
+    :returns: {(service, instance): [(hostname, container_id, process_name), ...] }
```
Comment (solarkennedy) [inline paasta_tools/check_oom_events.py]: These complicated return thingys are easy to get wrong. Can you implement this as a named tuple instead?

File: paasta_tools/check_oom_events.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    parser.add_argument(
+        '-s', '--superregion', dest="superregion", required=True,
+        help="The superregion to read OOM events from.",
+    )
+    return parser.parse_args(args)
+
+
+def oom_events(cluster, superregion, num_lines=1000):
+    """Read the latest 'num_lines' lines from OOM_EVENTS_STREAM and iterate over them."""
+    host_port = choice(scribereader.get_default_scribe_hosts(tail=True))
```
Comment (solarkennedy) [inline paasta_tools/check_oom_events.py]: whats up with this choice? Is this really the normal way to read scribe?

Comment (oktopuz) [inline paasta_tools/check_oom_events.py]: https://opengrok.yelpcorp.com/xref/z_everything_else/scribereader/scribereader/scribereader.py#105

get_default_scribe_hosts may return more than 1 host.

File: paasta_tools/check_oom_events.py (hunk 7 above)
Comment (oktopuz) [inline paasta_tools/check_oom_events.py]: I addressed everything except this one ^^, because this message is too logs for irc.

**Pull request merged**